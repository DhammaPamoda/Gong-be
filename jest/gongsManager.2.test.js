const fs = require('fs');
const moment = require('moment');
const gongsManager = require('../lib/gongsManager');
const logger = require('../lib/logger');

logger.setTestMode();

const dateFormat = 'YYYY-MM-DD';
const validCourseName = '10_day_course';

const jobActionStubFunction = jest.fn((/* Job */aJobOrJobs, aAction) => {
  // console.log('aaaa', aJobOrJobs.length, aAction);
});

describe('Test gongsManager courses hanling', () => {
  let courseDaysCount;

  beforeAll(() => {
    const filesPath = 'jest/data/1';
    fs.writeFileSync(`${filesPath}/coursesSchedule.json`, '[]');
    gongsManager.setDynamicJsonFilesPath(filesPath);
    gongsManager.addOnGongActionListener(jobActionStubFunction);
    gongsManager.init();
    jobActionStubFunction.mockClear();
  });

  const addCourseFailedFn = async (aCourseJson) => {
    let retValue;
    try {
      retValue = await gongsManager.addScheduledCourse(aCourseJson);
    } catch (e) {
      expect(e).toBeInstanceOf(Error);
      throw e;
    }
    return retValue;
  };

  const removeCourseFn = async (aCourse4RemoalJson) => {
    let retValue;
    try {
      retValue = await gongsManager.removeScheduledCourse(aCourse4RemoalJson);
    } catch (e) {
      throw e;
    }
    return retValue;
  };

  const removeCourseFailedTestFn = (aCourse4RemoalJson, aTestDescription) => {
    test(aTestDescription, async () => {
      const noOfCoursesBeforeRemove = gongsManager.scheduledCoursesArray.length;
      jobActionStubFunction.mockClear();
      expect.assertions(4);

      await expect(removeCourseFn(aCourse4RemoalJson)).rejects.toThrow();

      expect(jobActionStubFunction.mock.calls.length).toBe(0);
      expect(gongsManager.scheduledCoursesArray.length).toEqual(noOfCoursesBeforeRemove);
      expect(gongsManager.automaticGongsMap.size).toEqual(noOfCoursesBeforeRemove);
    });
  };

  const addCourseFailedTestFn = (aCourseJson, aTestDescription) => {
    test(aTestDescription, async () => {
      jobActionStubFunction.mockClear();
      expect.assertions(4);

      await expect(gongsManager.addScheduledCourse(aCourseJson)).rejects.toThrow();

      expect(jobActionStubFunction.mock.calls.length).toBe(0);
      expect(gongsManager.scheduledCoursesArray.length).toMatchSnapshot();
      expect(gongsManager.automaticGongsMap.size).toMatchSnapshot();
    });
  };

  describe('Failed courses addition ', () => {
    beforeAll(() => {
      const course = gongsManager.coursesMap.get(validCourseName);
      courseDaysCount = course.daysCount;
    });

    addCourseFailedTestFn(null, 'Null object');

    let scheduledCourseJson = {};
    addCourseFailedTestFn(scheduledCourseJson, 'Empty object');

    const validTimeForCourse = moment();
    scheduledCourseJson = {
      date: validTimeForCourse.format(dateFormat),
    };
    addCourseFailedTestFn(scheduledCourseJson, 'No Course name');

    scheduledCourseJson = {
      course_name: 'aaa',
    };
    addCourseFailedTestFn(scheduledCourseJson, 'No time');

    scheduledCourseJson = {
      course_name: validCourseName,
      date: 'aaaa',
    };
    addCourseFailedTestFn(scheduledCourseJson, 'Invalid time');

    scheduledCourseJson = {
      course_name: validCourseName,
      date: validTimeForCourse.clone().subtract(1, 'y').format(dateFormat),
    };
    addCourseFailedTestFn(scheduledCourseJson, 'Past time');

  });

  test('Valid courses addition', async () => {
    const currentMoment = moment();
    const dateStr = currentMoment.format(dateFormat);
    const scheduledCourseJson = {
      course_name: validCourseName,
      date: dateStr,
    };

    const addScheduledCourseSuccess = await gongsManager.addScheduledCourse(scheduledCourseJson);
    expect(addScheduledCourseSuccess).toMatchSnapshot({
      id: expect.any(Number),
      date: expect.any(String),
    });
    expect(jobActionStubFunction.mock.calls.length).toMatchSnapshot();
    expect(gongsManager.scheduledCoursesArray.length).toEqual(1);
    expect(gongsManager.automaticGongsMap.size).toEqual(1);
  });

  describe('Check for collision ', () => {
    let validTimeForCourse;
    let scheduledCourseJson;
    beforeAll(() => {
      validTimeForCourse = moment();
    });
    beforeEach(() => jobActionStubFunction.mockClear());

    test('Adding/Subtracting days - check for fail ', async () => {
      expect.assertions(27);

      for (let i = 0; i < courseDaysCount; i++) {
        scheduledCourseJson = {
          course_name: validCourseName,
          date: validTimeForCourse.clone().add(i, 'd').format(dateFormat),
        };
        await expect(gongsManager.addScheduledCourse(scheduledCourseJson)).rejects.toThrow();

        scheduledCourseJson = {
          course_name: validCourseName,
          date: validTimeForCourse.clone().subtract(i, 'd').format(dateFormat),
        };
        await expect(gongsManager.addScheduledCourse(scheduledCourseJson)).rejects.toThrow();
      }
      expect(jobActionStubFunction.mock.calls.length).toBe(0);
      expect(gongsManager.scheduledCoursesArray.length).toMatchSnapshot();
      expect(gongsManager.automaticGongsMap.size).toMatchSnapshot();
    });

    test('Subtracting days - check for failure - before current time ', async () => {
      expect.assertions(4);
      scheduledCourseJson = {
        course_name: validCourseName,
        date: validTimeForCourse.clone().subtract(courseDaysCount, 'd').format(dateFormat),
      };
      await expect(gongsManager.addScheduledCourse(scheduledCourseJson)).rejects.toThrow();

      expect(jobActionStubFunction.mock.calls.length).toBe(0);
      expect(gongsManager.scheduledCoursesArray.length).toMatchSnapshot();
      expect(gongsManager.automaticGongsMap.size).toMatchSnapshot();
    });

    test('Adding days - check for success ', async () => {
      scheduledCourseJson = {
        course_name: validCourseName,
        date: validTimeForCourse.clone().add(courseDaysCount, 'd').format(dateFormat),
      };
      const addedCourse = await addCourseFailedFn(scheduledCourseJson);
      expect(addedCourse).toBeTruthy();

      expect(jobActionStubFunction.mock.calls.length).toBeGreaterThan(0);
      expect(jobActionStubFunction.mock.calls.length).toMatchSnapshot();
      expect(gongsManager.scheduledCoursesArray.length).toMatchSnapshot();
      expect(gongsManager.automaticGongsMap.size).toMatchSnapshot();
    });
  });

  describe('Removing courses ', () => {
    let validTimeForCourse;
    let scheduledCourse4RemovalJson;
    beforeAll(() => {
      validTimeForCourse = moment();
    });
    beforeEach(() => jobActionStubFunction.mockClear());

    describe('Failed removing courses ', () => {
      removeCourseFailedTestFn(null, 'Null object');
      scheduledCourse4RemovalJson = {};
      removeCourseFailedTestFn(scheduledCourse4RemovalJson, 'Empty object');
      scheduledCourse4RemovalJson = {
        id: 'aaaa',
      };
      removeCourseFailedTestFn(scheduledCourse4RemovalJson, 'Invalid id');
      scheduledCourse4RemovalJson = {
        id: 3333,
      };
      removeCourseFailedTestFn(scheduledCourse4RemovalJson, 'Id that do not exists');
    });

    test('Successed removing courses ', async () => {
      const noOfCoursesBeforeRemove = gongsManager.scheduledCoursesArray.length;
      scheduledCourse4RemovalJson = {
        id: moment().startOf('day').valueOf(),
      };
      const removeResult = await removeCourseFn(scheduledCourse4RemovalJson)
      expect(removeResult).toEqual('SUCCESS');
      expect(gongsManager.scheduledCoursesArray.length).toEqual(noOfCoursesBeforeRemove - 1);
      expect(gongsManager.automaticGongsMap.size).toEqual(noOfCoursesBeforeRemove - 1);
    });
  });


});
