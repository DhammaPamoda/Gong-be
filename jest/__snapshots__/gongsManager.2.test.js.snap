// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test gongsManager courses hanling Check for collision  Adding days - check for success  1`] = `94`;

exports[`Test gongsManager courses hanling Check for collision  Adding days - check for success  2`] = `2`;

exports[`Test gongsManager courses hanling Check for collision  Adding days - check for success  3`] = `2`;

exports[`Test gongsManager courses hanling Check for collision  Adding days - check for success  4`] = `2`;

exports[`Test gongsManager courses hanling Check for collision  Adding days - check for success  5`] = `2`;

exports[`Test gongsManager courses hanling Check for collision  Adding/Subtracting days - check for fail  1`] = `1`;

exports[`Test gongsManager courses hanling Check for collision  Adding/Subtracting days - check for fail  2`] = `1`;

exports[`Test gongsManager courses hanling Check for collision  Subtracting days - check for failure - before current time  1`] = `1`;

exports[`Test gongsManager courses hanling Check for collision  Subtracting days - check for failure - before current time  2`] = `1`;

exports[`Test gongsManager courses hanling Failed courses addition  Empty object 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Empty object 2`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Invalid time 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Invalid time 2`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  No Course name 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  No Course name 2`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  No time 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  No time 2`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Null object 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Null object 2`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Past time 1`] = `0`;

exports[`Test gongsManager courses hanling Failed courses addition  Past time 2`] = `0`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Empty object 1`] = `1`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Empty object 2`] = `2`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Id that do not exists 1`] = `0`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Id that do not exists 2`] = `2`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Invalid id 1`] = `0`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Invalid id 2`] = `2`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Null object 1`] = `2`;

exports[`Test gongsManager courses hanling Removing courses  Removing courses  Null object 2`] = `2`;

exports[`Test gongsManager courses hanling Valid courses addition 1`] = `
Object {
  "course_name": "10_day_course",
  "date": Any<String>,
  "id": Any<Number>,
  "startFromDay": undefined,
  "testHoursRange": undefined,
}
`;

exports[`Test gongsManager courses hanling Valid courses addition 2`] = `94`;
